analysis/bulk-deconvolution/scripts/calculate-tpm.R

```diff
@@ -0,0 +1,124 @@
+  ),
+  make_option(
+    "--output_dir",
+    type = "character",
+    help = "Output directory to save TSV with TPM values."
+  ),
+  make_option(
+    "--t2g_file",
+    type = "character",
+    default = here::here("s3_files", "reference_files", "Homo_sapiens.GRCh38.104.spliced_cdna.tx2gene.tsv"),
+    help = "Path to the t2g file used to read in quant.sf files."
+  ),
+  make_option(
+    "--scpca_sce_file",
+    type = "character",
+    default = here::here("s3_files", "SCPCS000001", "SCPCL000001_processed.rds"),
+    help = "Path to an SCE file used to obtain mappings for converting ensembl ids to gene symbols."
+  )
+)
+opts <- parse_args(OptionParser(option_list = option_list))
+
+# Check inputs and define paths -------
+stopifnot(
+  "A project id in the format SCPCPXXXXXX must be provided to `project_id`." = !is.null(opts$project_id),
+  "An output directory must be specified with `output_dir`." = !is.null(opts$output_dir)
+)
+
+data_dir <- file.path(opts$input_dir, opts$project_id)
+stopifnot("Provided `input_dir` does not contain a directory with quant files for the specified `project_id`." = dir.exists(data_dir))
+
+fs::dir_create(opts$output_dir)
+output_file <- file.path(opts$output_dir, glue::glue("{opts$project_id}-tpm.tsv"))
+
+stopifnot("The t2g file could not be found." = file.exists(opts$t2g_file))
+stopifnot("The SCE file to use for id mapping could not be found." = file.exists(opts$scpca_sce_file))
+
+
+# Read input helper files -----------
+t2g_table <- readr::read_tsv(
+  opts$t2g_file,
+  col_names = c("transcript_id", "gene_id"),
+  show_col_types = FALSE
+)
+sce <- readRDS(opts$scpca_sce_file)
+map_table <- data.frame(
+  ensembl_id = SingleCellExperiment::rowData(sce)$gene_ids,
+  gene_symbol = SingleCellExperiment::rowData(sce)$gene_symbol
+)
+
+
+# Find all quant.sf files -------------
+quant_files <- list.files(
+  path = data_dir,
+  recursive = TRUE
+)
+stopifnot("Could not find any quant.sf files for the specified project." = length(quant_files) > 0)
+
+# Calculate TPM for each sample and combine into single data frame ---------
+tpm_df <- quant_files |>
+  purrr::map(
+    \(quant_file_path) {
+
+      split_file <- stringr::str_split_1(quant_file_path, pattern = "/")
+      sample_id <- split_file[1]
+      library_id <- split_file[2]
+
+      # read in with tximport and extract TPM
+      txi_salmon <- tximport::tximport(
+        file.path(data_dir, quant_file_path),
+        type = "salmon",
+        tx2gene = t2g_table
+      )
+      tpm <- txi_salmon$abundance[,1]
+      ensembl_ids <- names(tpm)
+
+      # return data frame where the tpm column is named by the sample id
+      sample_tpm_df <- data.frame(tpm = tpm)
+      names(sample_tpm_df) <- sample_id
+      sample_tpm_df
+
+    }
+  ) |>
+  # we expect the same genes all around here since quantification used the same gene set
+  dplyr::bind_cols() |>
+  tibble::rownames_to_column(var = "ensembl_id")
```

Comment:
You can let `tximport` do all of the table making here: you just need `quant_files` to be named with the sample_ids
```suggestion

quant_names <- stringr::str_split_i(quant_files, pattern = "/", i = 2)
quant_paths <- setNames(file.path(data_dir, quant_files), quant_names)


txi_salmon <- tximport::tximport(
  quant_paths,
  type = "salmon",
  tx2gene = t2g_table
)

tpm_mat  <- txi_salmon$abundance
```

Then `rowsum()` can replace all of the tidyverse stuff:

```
symbol_map <- setNames(map_table$gene_symbol, map_table$ensembl_id)
tpm_summed_mat <- rowsum(
  tpm_mat,
  group = symbol_map[rownames(tpm_mat)]
)
```


The only other thing I might suggest is that you _could_ avoid loading the tx2gene info by using `tximeta` but since we already have all that done, there isn't too much point. Though it should get you the symbol translations too, without having to load a whole SCE object.

Reply:
>  But you might move getting the translation table from an SCE to the previous script, just to prevent loading the whole SCE for each project.

And export a TSV of the "map_table" to read in here, you mean?